Cython/Compiler/ExprNodes.py

```diff
@@ -8787,6 +8790,79 @@ def generate_evaluation_code(self, code):
             arg.free_temps(code)
 
 
+class FrozenSetNode(SetNode):
+    """
+    Frozenset constructor.
+    """
+
+    gil_message = "Constructing Python frozenset"
+    is_literal = False
+    type = frozenset_type
+
+    def calculate_result_code(self):
+        return self.result_code
+
+    def _generate_creating_from_constant_code(self, code, target):
+        if target is None:
+            target = self.result()
+        assert target is not None
+        if self.args is None:
+            constructor_parameter = 0
+        else:
+            self.args.generate_evaluation_code(code)
+            constructor_parameter = self.args.py_result()
+        # TODO Avoid code duplication for creating a frozenset
+        code.globalstate.use_utility_code(UtilityCode.load_cached(
+            'pyfrozenset_new', 'Builtins.c'))
+        code.putln("%s = __Pyx_PyFrozenSet_New(%s); %s" % (
+            target, constructor_parameter,
+            code.error_goto_if_null(target, self.pos)))
+        self.result_code = target
+
+# TODO: for such case
+#     x = frozenset((1, 2, 3, 2))
+#     y = frozenset((1, 2, 3, 2, 3, 1))
+#     _get_dedup_values will generate same dedup key,
+#     but cython will still try to create frozenset from a tuple with duplicate elements
+
+    def _get_dedup_values(self, args):
+        if args is None:
+            return ()
+        return args.args
+
+    def _create_shared_frozenset_object(self, code):
+        assert self.type.name == 'frozenset'
+        dedup_key = make_dedup_key(self.type, self._get_dedup_values(self.args))
+        set_target = code.get_py_const(py_object_type, 'frozenset', cleanup_level=2, dedup_key=dedup_key)
+        assert set_target is not None
+        const_code = code.get_cached_constants_writer(set_target)
+        if const_code is not None:
+            # constant is not yet initialised
+            const_code.mark_pos(self.pos)
+            self._generate_creating_from_constant_code(const_code, set_target)
+        self.result_code = set_target
+
+    def generate_evaluation_code(self, code):
+        print(self.__dict__)
+        if self.is_literal:
+            self._create_shared_frozenset_object(code)
+            return
+
+        self.args.generate_evaluation_code(code)
+        self.allocate_temp_result(code)
+        code.globalstate.use_utility_code(UtilityCode.load_cached(
+            'pyfrozenset_new', 'Builtins.c'))
+
+        code.putln(
+            "%s = __Pyx_PyFrozenSet_New(%s); %s" % (
+                self.result(),
+                self.args.py_result(),
+                code.error_goto_if_null(self.result(), self.pos)))
+        self.generate_gotref(code)
+        self.args.generate_disposal_code(code)
+        self.args.free_temps(code)
```

Comment:
This code seems a (more correct) duplicate of the code in `_generate_creating_from_constant_code()`. I think it's worth using the same implementation for both.

Reply:
I agree this code duplication should be eliminated.  I'm thinking about it now